### DISCUSSION TITLE
OTBR cannot switch to Leader

### DISCUSSION THREAD

**Discussion by aaa147147** [openthread/openthread#11019] [General]
**Describe the bug** A clear and concise description of what the bug is.
OTBR cannot switch to Leader -> Error 27: NotCapable
**To Reproduce** Information to reproduce the behavior, including:

1. Git commit id: ->Commit 147de7e
2. IEEE 802.15.4 hardware platform -> RCP: RTL8771HTV  SOC: Amlogic S905X5M
4. Build steps -> Attachment<options.cmake.txt>
5. Network topology -> I don't know how to describe it, can you give me more specific tips

**Expected behavior** A clear and concise description of what you expected to happen.
OTBR switches to Leader normally

**Console/log output** If applicable, add console/log output to help explain your problem.
We conducted a power on/off test, and the log showed that during the 88th power on/off test, the status could not be switched to Leader
The 89th test log we have attached can be switched to Leader normally
![image](https://github.com/user-attachments/assets/e0624dd5-d23c-4492-8942-26e1dbbb92d0)

**Additional context** Add any other context about the problem here.
 Attachment<88.log>   -> Error
 Attachment<89.log>   -> OK

**Comment by aaa147147** [comment]
[88.log](https://github.com/user-attachments/files/18033310/88.log)
[89.log](https://github.com/user-attachments/files/18033311/89.log)
[options.cmake.txt](https://github.com/user-attachments/files/18033324/options.cmake.txt)

**Comment by aaa147147** [comment]
Displayed in<995. log>, unable to switch to leader, and meshdiag topology children ip6 addrs returned timeout
![image](https://github.com/user-attachments/assets/b88027e7-649f-4899-82d5-e5bc4cab68e7)
[995.log](https://github.com/user-attachments/files/18033464/995.log)

**Comment by jwhui** [comment]
The `state leader` command calls `Mle::BecomeLeader()`. The `NotCapable` error is returned in the following cases: 
https://github.com/openthread/openthread/blob/7b72e9778253a87c404cc27af85be5cf12b0ef11/src/core/thread/mle_router.cpp#L281
https://github.com/openthread/openthread/blob/7b72e9778253a87c404cc27af85be5cf12b0ef11/src/core/thread/mle_router.cpp#L285

I suspect the error is due to the second case, which indicates that the device is already attached.

**Comment by aaa147147** [comment]
@jwhui Thanks for your reply
Our current approach is to first change the leaderweight to 255, then state leader for switch the role to leader
![image](https://github.com/user-attachments/assets/33d1edfa-f856-492c-a180-19596d3f7734)
So, **what checks do we need to do** before switching roles to leaders to ensure that we can switch roles to leaders

**Comment by jwhui** [comment]
It would be good to check whether the device is already attached and, if attached, check the leader weight of the partition it is currently attached to.

**Comment by aaa147147** [comment]
@jwhui Okay, in the next test, before modifying the leader weight, we will use the `leaderweight ` `state` commands to check the current state

**Comment by aaa147147** [comment]
@jwhui In fact, obtaining the connection status and leader weight value is only for our next step of problem localization. 

For example, when the state is child and the leader weight is 64, we cannot switch to leader, but first switch to router and then switch to leader

Is there some similar requirements?

**Reply by jwhui** [reply]
I don't think so. See the `BecomeLeader()` implementation the checks that are in place:

https://github.com/openthread/openthread/blob/7b72e9778253a87c404cc27af85be5cf12b0ef11/src/core/thread/mle_router.cpp#L267
